lib/async_hooks.js

```diff
@@ -98,24 +103,19 @@ class AsyncHook {
     if (hooks_array.includes(this))
       return;
 
-    if (!setupHooksCalled) {
-      setupHooksCalled = true;
-      // Setup the callbacks that node::AsyncWrap will call when there are
-      // hooks to process. They use the same functions as the JS embedder API.
-      async_wrap.setupHooks({ init,
-                              before: emitBeforeN,
-                              after: emitAfterN,
-                              destroy: emitDestroyN });
-    }
-
     // createHook() has already enforced that the callbacks are all functions,
     // so here simply increment the count of whether each callbacks exists or
     // not.
     hook_fields[kInit] += +!!this[init_symbol];
     hook_fields[kBefore] += +!!this[before_symbol];
     hook_fields[kAfter] += +!!this[after_symbol];
     hook_fields[kDestroy] += +!!this[destroy_symbol];
+    hook_fields[kTotals]++;
```

Comment:
Sorry, reading back over it the comment wasn't correct. I was thinking about the case where async hooks are enabled in the middle of executing a promise chain and the parent promise is not wrapped but I forgot this was handled by adding a silent mode wrap when the child promise is created (https://github.com/trevnorris/node/blob/5a61e784c9b4f03a1341970a0b81819a52a60dde/src/async-wrap.cc#L309). However, the first part of the comment still holds. I was curious whether having async hooks enabled with no callbacks registered was expected to occur frequently enough to make it worth optimizing?